### DISCUSSION TITLE
TailwindSort on save

### DISCUSSION THREAD

**Discussion by Gnarus-G** [luckasRanarison/tailwind-tools.nvim#15] [Q&A] [answered]
I want to be able to run TailwindSort on save, but the sorting procedure is asynchronous so it's not guaranteed to resolve before the buffer is written. Not a big deal for me - so not ready to make a feature request - but would be nice if the api could easily handle it.

**Comment by luckasRanarison** [comment]
Yeah, I deliberately didn't add sort on save because of the asynchronous nature of the function, but I'm confused what kind of API do you want?

**Reply by Gnarus-G** [reply]
It's not that there's a kind of API that I want. I was wondering if there was a way to get this feature, despite the async nature of it, in a way that doesn't break the current API. The current API being the setup function with the options table.

I should note that I'm only asking about this out of wishful thinking. Based on everything I know I would have also not added sort on save. I was just wondering if there was a worthwhile workaround that I didn't know.

Simply put, I am satisfied with the way things work now because I think the alternative might either create a bad API or bad internal hacky code.

**Reply by luckasRanarison** [reply, marked as answer]
The problem with sorting on save is that we don't know how the users handle BufWrite events, do they use format on save? If so is it asynchronous or not? And in the case where both are async we need a callback API on one side, the default `format` function from neovim doesn't provide a callback argument so I think it's resonable to a callback function on our side or add a sync version (I personally prefer async), but currently there's no way to do that if both are async.

**Comment by tom-gora** [comment]
Been a while but since I just solved this for myself I thought I'll share if anyone wants to get sorting to work.

_Prerequisites: using conform.nvim for managing formatters or I guess writing own autocmd logic_
as per [conform's docs](https://github.com/stevearc/conform.nvim/blob/master/doc/recipes.md#autoformat-with-extra-features):
`-- There is a similar affordance for format_after_save, which uses BufWritePost.
-- This is good for formatters that are too slow to run synchronously.`

I placed the following key in my conform's opts:
```lua
	format_after_save = function()
		local t_attached = vim.tbl_contains(
			vim.tbl_map(function(c)
				return c.name
			end, vim.lsp.get_clients()),
			"tailwindcss"
		)
		if not t_attached or not pcall(require, "tailwind-tools") then
			return
		end

		vim.cmd("TailwindSort")
		-- These options will be passed to conform.format()
		return { lsp_format = "fallback" }
	end,
```
and it runs the format for me just fine.

**Reply by luckasRanarison** [reply]
Nice, thanks for sharing! There's also `TailwindSortSync` now, just in case other people want a blocking version of it.
